File: modules/vescape-core/android/src/main/java/expo/modules/vescapecore/config/RefloatConfigProtocol.kt
package expo.modules.vescapecore.config

import expo.modules.vescapecore.connection.BoardTransport
import expo.modules.vescapecore.protocol.COMM_CUSTOM_APP_DATA
import expo.modules.vescapecore.protocol.COMM_FORWARD_CAN
import expo.modules.vescapecore.protocol.COMM_GET_CUSTOM_CONFIG
import expo.modules.vescapecore.protocol.COMM_GET_CUSTOM_CONFIG_XML
import expo.modules.vescapecore.protocol.COMM_SET_CUSTOM_CONFIG
import expo.modules.vescapecore.protocol.REFLOAT_GET_INFO
import expo.modules.vescapecore.protocol.REFLOAT_MAGIC

import java.nio.ByteBuffer
import java.nio.ByteOrder

// @parity /modules/vescape-core/ios/config/RefloatConfigProtocol.swift
internal data class RefloatConfigXmlChunk(
  val confInd: Int,
  val totalLength: Int,
  val offset: Int,
  val chunk: ByteArray,
)

internal data class RefloatConfigBytes(
  val confInd: Int,
  val packageSignature: Long,
  val config: ByteArray,
)

internal data class RefloatPackageInfo(
  val version: String,
)

internal sealed class RefloatConfigProtocolResult<out T> {
  data class Success<T>(val value: T) : RefloatConfigProtocolResult<T>()
  data class Failure(val message: String) : RefloatConfigProtocolResult<Nothing>()
}

internal object RefloatConfigProtocol {
  private val baseVersionPattern = Regex("""\b(\d+\.\d+(?:\.\d+)?)\b""")

  fun normalizeBaseVersion(version: String?): String? =
    version
      ?.trim()
      ?.takeIf { it.isNotEmpty() }
      ?.let { baseVersionPattern.find(it)?.groupValues?.get(1) }

  private fun commandOffset(payload: ByteArray, expectedCommand: Int): RefloatConfigProtocolResult<Int> {
    if (payload.isEmpty()) {
      return RefloatConfigProtocolResult.Failure("Empty Refloat config response")
    }
    val cmd = payload[0].toInt() and 0xff
    if (cmd == expectedCommand) return RefloatConfigProtocolResult.Success(0)
    if (cmd == COMM_FORWARD_CAN) {
      if (payload.size < 3) {
        return RefloatConfigProtocolResult.Failure("Short forwarded Refloat config response")
      }
      val forwarded = payload[2].toInt() and 0xff
      if (forwarded == expectedCommand) return RefloatConfigProtocolResult.Success(2)
      return RefloatConfigProtocolResult.Failure(
        "Unexpected forwarded Refloat config command $forwarded, expected $expectedCommand",
      )
    }
    return RefloatConfigProtocolResult.Failure(
      "Unexpected Refloat config command $cmd, expected $expectedCommand",
    )
  }

  private fun appCommandOffset(payload: ByteArray, expectedCommand: Int): RefloatConfigProtocolResult<Int> {
    if (payload.size < 3) {
      return RefloatConfigProtocolResult.Failure("Short Refloat app response")
    }
    val cmd = payload[0].toInt() and 0xff
    if (cmd == COMM_CUSTOM_APP_DATA) {
      val magic = payload[1].toInt() and 0xff
      val appCommand = payload[2].toInt() and 0xff
      if (magic != REFLOAT_MAGIC) {
        return RefloatConfigProtocolResult.Failure("Unexpected Refloat magic $magic")
      }
      if (appCommand == expectedCommand) return RefloatConfigProtocolResult.Success(2)
      return RefloatConfigProtocolResult.Failure(
        "Unexpected Refloat app command $appCommand, expected $expectedCommand",
      )
    }
    if (cmd == COMM_FORWARD_CAN) {
      if (payload.size < 5) {
        return RefloatConfigProtocolResult.Failure("Short forwarded Refloat app response")
      }
      val forwarded = payload[2].toInt() and 0xff
      val magic = payload[3].toInt() and 0xff
      val appCommand = payload[4].toInt() and 0xff
      if (forwarded != COMM_CUSTOM_APP_DATA) {
        return RefloatConfigProtocolResult.Failure(
          "Unexpected forwarded Refloat command $forwarded, expected $COMM_CUSTOM_APP_DATA",
        )
      }
      if (magic != REFLOAT_MAGIC) {
        return RefloatConfigProtocolResult.Failure("Unexpected Refloat magic $magic")
      }
      if (appCommand == expectedCommand) return RefloatConfigProtocolResult.Success(4)
      return RefloatConfigProtocolResult.Failure(
        "Unexpected forwarded Refloat app command $appCommand, expected $expectedCommand",
      )
    }
    return RefloatConfigProtocolResult.Failure(
      "Unexpected Refloat app response command $cmd, expected $COMM_CUSTOM_APP_DATA",
    )
  }

  fun buildGetInfo(transport: BoardTransport, version: Int = 1): ByteArray {
    require(version in 0..255) { "version must fit uint8" }
    return transport.frame(
      byteArrayOf(
        COMM_CUSTOM_APP_DATA.toByte(),
        REFLOAT_MAGIC.toByte(),
        REFLOAT_GET_INFO.toByte(),
        version.toByte(),
      ),
    )
  }

  fun buildGetCustomConfigXml(
    transport: BoardTransport,
    confInd: Int,
    length: Int,
    offset: Int,
  ): ByteArray {
    require(confInd in 0..255) { "confInd must fit uint8" }
    require(length >= 0) { "length must be non-negative" }
    require(offset >= 0) { "offset must be non-negative" }
    val cmd = ByteBuffer.allocate(10)
      .order(ByteOrder.BIG_ENDIAN)
      .put(COMM_GET_CUSTOM_CONFIG_XML.toByte())
      .put(confInd.toByte())
      .putInt(length)
      .putInt(offset)
      .array()
    return transport.frame(cmd)
  }

  fun buildGetCustomConfig(transport: BoardTransport, confInd: Int): ByteArray {
    require(confInd in 0..255) { "confInd must fit uint8" }
    return transport.frame(byteArrayOf(COMM_GET_CUSTOM_CONFIG.toByte(), confInd.toByte()))
  }

  fun parseCustomConfigXmlResponse(
    payload: ByteArray,
    expectedConfInd: Int = 0,
  ): RefloatConfigProtocolResult<RefloatConfigXmlChunk> {
    val cmdOffset = when (val result = commandOffset(payload, COMM_GET_CUSTOM_CONFIG_XML)) {
      is RefloatConfigProtocolResult.Success -> result.value
      is RefloatConfigProtocolResult.Failure -> return result
    }
    if (payload.size < cmdOffset + 10) {
      return RefloatConfigProtocolResult.Failure(
        "Short Refloat config XML response: ${payload.size - cmdOffset} bytes",
      )
    }
    val view = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
    view.position(cmdOffset + 1)
    val confInd = view.get().toInt() and 0xff
    if (confInd != expectedConfInd) {
      return RefloatConfigProtocolResult.Failure("Unexpected Refloat config XML index $confInd")
    }
    val totalLength = view.int
    val dataOffset = view.int
    if (totalLength < 0) {
      return RefloatConfigProtocolResult.Failure("Negative Refloat config XML length $totalLength")
    }
    if (dataOffset < 0 || dataOffset > totalLength) {
      return RefloatConfigProtocolResult.Failure(
        "Invalid Refloat config XML offset $dataOffset for length $totalLength",
      )
    }
    val chunk = payload.copyOfRange(cmdOffset + 10, payload.size)
    if (dataOffset + chunk.size > totalLength) {
      return RefloatConfigProtocolResult.Failure(
        "Refloat config XML chunk exceeds length: offset=$dataOffset chunk=${chunk.size} length=$totalLength",
      )
    }
    return RefloatConfigProtocolResult.Success(RefloatConfigXmlChunk(confInd, totalLength, dataOffset, chunk))
  }

  fun buildSetCustomConfig(
    transport: BoardTransport,
    confInd: Int,
    packageSignature: Long,
    configBytes: ByteArray,
  ): ByteArray {
    require(confInd in 0..255) { "confInd must fit uint8" }
    val buf = ByteBuffer.allocate(6 + configBytes.size).order(ByteOrder.BIG_ENDIAN)
    buf.put(COMM_SET_CUSTOM_CONFIG.toByte())
    buf.put(confInd.toByte())
    buf.putInt(packageSignature.toInt())
    buf.put(configBytes)
    return transport.frame(buf.array())
  }

  fun parseSetCustomConfigResponse(
    payload: ByteArray,
    expectedConfInd: Int = 0,
  ): RefloatConfigProtocolResult<Int> {
    val offset = when (val result = commandOffset(payload, COMM_SET_CUSTOM_CONFIG)) {
      is RefloatConfigProtocolResult.Success -> result.value
      is RefloatConfigProtocolResult.Failure -> return result
    }
    if (payload.size == offset + 1) {
      return RefloatConfigProtocolResult.Success(expectedConfInd)
    }
    val confInd = payload[offset + 1].toInt() and 0xff
    if (confInd != expectedConfInd) {
      return RefloatConfigProtocolResult.Failure("Unexpected Refloat set config index $confInd")
    }
    return RefloatConfigProtocolResult.Success(confInd)
  }

  fun parseCustomConfigResponse(
    payload: ByteArray,
    expectedConfInd: Int = 0,
  ): RefloatConfigProtocolResult<RefloatConfigBytes> {
    val offset = when (val result = commandOffset(payload, COMM_GET_CUSTOM_CONFIG)) {
      is RefloatConfigProtocolResult.Success -> result.value
      is RefloatConfigProtocolResult.Failure -> return result
    }
    if (payload.size < offset + 6) {
      return RefloatConfigProtocolResult.Failure(
        "Short Refloat config response: ${payload.size - offset} bytes",
      )
    }
    val view = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
    view.position(offset + 2)
    val confInd = payload[offset + 1].toInt() and 0xff
    if (confInd != expectedConfInd) {
      return RefloatConfigProtocolResult.Failure("Unexpected Refloat config index $confInd")
    }
    val packageSignature = view.int.toLong() and 0xffffffffL
    return RefloatConfigProtocolResult.Success(
      RefloatConfigBytes(
        confInd = confInd,
        packageSignature = packageSignature,
        config = payload.copyOfRange(offset + 6, payload.size),
      ),
    )
  }

  fun parseGetInfoResponse(payload: ByteArray): RefloatConfigProtocolResult<RefloatPackageInfo> {
    val offset = when (val result = appCommandOffset(payload, REFLOAT_GET_INFO)) {
      is RefloatConfigProtocolResult.Success -> result.value
      is RefloatConfigProtocolResult.Failure -> return result
    }
    val dataOffset = offset + 1
    if (payload.size <= dataOffset) {
      return RefloatConfigProtocolResult.Failure("Short Refloat info response: 0 bytes")
    }
    val first = payload[dataOffset].toInt() and 0xff
    if (first == 2) return parseGetInfoV2(payload, dataOffset)
    return parseGetInfoV1(payload, dataOffset)
  }

  private fun parseGetInfoV1(
    payload: ByteArray,
    dataOffset: Int,
  ): RefloatConfigProtocolResult<RefloatPackageInfo> {
    if (payload.size < dataOffset + 3) {
      return RefloatConfigProtocolResult.Failure(
        "Short Refloat info v1 response: ${payload.size - dataOffset} bytes",
      )
    }
    val versionCode = payload[dataOffset].toInt() and 0xff
    val major = versionCode / 10
    val minor = versionCode % 10
    return RefloatConfigProtocolResult.Success(RefloatPackageInfo("Refloat $major.$minor"))
  }

  private fun parseGetInfoV2(
    payload: ByteArray,
    dataOffset: Int,
  ): RefloatConfigProtocolResult<RefloatPackageInfo> {
    val minLength = dataOffset + 2 + 20 + 3
    if (payload.size < minLength) {
      return RefloatConfigProtocolResult.Failure(
        "Short Refloat info v2 response: ${payload.size - dataOffset} bytes",
      )
    }
    val packageName = displayName(fixedString(payload, dataOffset + 2, 20))
    val major = payload[dataOffset + 22].toInt() and 0xff
    val minor = payload[dataOffset + 23].toInt() and 0xff
    val patch = payload[dataOffset + 24].toInt() and 0xff
    val suffix = if (payload.size >= dataOffset + 45) fixedString(payload, dataOffset + 25, 20) else ""
    val suffixPart = when {
      suffix.isBlank() -> ""
      suffix.startsWith("-") -> suffix
      else -> "-$suffix"
    }
    return RefloatConfigProtocolResult.Success(
      RefloatPackageInfo("$packageName $major.$minor.$patch$suffixPart"),
    )
  }

  private fun fixedString(payload: ByteArray, offset: Int, length: Int): String {
    val end = (offset until (offset + length).coerceAtMost(payload.size))
      .firstOrNull { payload[it] == 0.toByte() }
      ?: (offset + length).coerceAtMost(payload.size)
    return if (end <= offset) "" else String(payload, offset, end - offset, Charsets.UTF_8).trim()
  }

  private fun displayName(raw: String): String {
    if (raw.isBlank()) return "Refloat"
    return raw.substring(0, 1).uppercase() + raw.substring(1)
  }
}

File: modules/vescape-core/ios/config/RefloatConfigProtocol.swift
import Foundation

struct RefloatConfigXmlChunk {
  let confInd: Int
  let totalLength: Int
  let offset: Int
  let chunk: [UInt8]
}

struct RefloatConfigBytes {
  let confInd: Int
  let packageSignature: UInt32
  let config: [UInt8]
}

struct RefloatPackageInfo {
  let version: String
}

enum RefloatConfigProtocolResult<T> {
  case success(T)
  case failure(String)
}

/// COMM_GET_CUSTOM_CONFIG_XML / COMM_GET_CUSTOM_CONFIG / COMM_SET_CUSTOM_CONFIG framing, parity with Android.
///
/// @parity /modules/vescape-core/android/src/main/java/expo/modules/vescapecore/config/RefloatConfigProtocol.kt
enum RefloatConfigProtocol {
  static func normalizeBaseVersion(_ version: String?) -> String? {
    guard let version = version?.trimmingCharacters(in: .whitespacesAndNewlines), !version.isEmpty else {
      return nil
    }
    guard let match = version.range(of: #"\b\d+\.\d+(?:\.\d+)?\b"#, options: .regularExpression) else {
      return nil
    }
    return String(version[match])
  }

  static func buildGetInfo(transport: BoardTransport, version: Int = 1) -> [UInt8] {
    precondition((0...255).contains(version), "version must fit uint8")
    return transport.frame([
      UInt8(COMM_CUSTOM_APP_DATA),
      UInt8(REFLOAT_MAGIC),
      UInt8(REFLOAT_GET_INFO),
      UInt8(version),
    ])
  }

  static func buildGetCustomConfigXml(
    transport: BoardTransport,
    confInd: Int,
    length: Int,
    offset: Int
  ) -> [UInt8] {
    precondition((0...255).contains(confInd), "confInd must fit uint8")
    precondition(length >= 0, "length must be non-negative")
    precondition(offset >= 0, "offset must be non-negative")
    return transport.frame([
      UInt8(COMM_GET_CUSTOM_CONFIG_XML),
      UInt8(confInd),
    ] + int32Bytes(length) + int32Bytes(offset))
  }

  static func buildGetCustomConfig(transport: BoardTransport, confInd: Int) -> [UInt8] {
    precondition((0...255).contains(confInd), "confInd must fit uint8")
    return transport.frame([UInt8(COMM_GET_CUSTOM_CONFIG), UInt8(confInd)])
  }

  static func buildSetCustomConfig(
    transport: BoardTransport,
    confInd: Int,
    packageSignature: UInt32,
    configBytes: [UInt8]
  ) -> [UInt8] {
    precondition((0...255).contains(confInd), "confInd must fit uint8")
    var frame: [UInt8] = [UInt8(COMM_SET_CUSTOM_CONFIG), UInt8(confInd)]
    frame.append(UInt8((packageSignature >> 24) & 0xff))
    frame.append(UInt8((packageSignature >> 16) & 0xff))
    frame.append(UInt8((packageSignature >> 8) & 0xff))
    frame.append(UInt8(packageSignature & 0xff))
    frame.append(contentsOf: configBytes)
    return transport.frame(frame)
  }

  static func parseSetCustomConfigResponse(
    _ payload: [UInt8],
    expectedConfInd: Int = 0
  ) -> RefloatConfigProtocolResult<Int> {
    let offset: Int
    switch commandOffset(payload, expectedCommand: COMM_SET_CUSTOM_CONFIG) {
    case .success(let value): offset = value
    case .failure(let message): return .failure(message)
    }
    if payload.count == offset + 1 {
      return .success(expectedConfInd)
    }
    let confInd = Int(payload[offset + 1])
    if confInd != expectedConfInd {
      return .failure("Unexpected Refloat set config index \(confInd)")
    }
    return .success(confInd)
  }

  static func parseCustomConfigXmlResponse(
    _ payload: [UInt8],
    expectedConfInd: Int = 0
  ) -> RefloatConfigProtocolResult<RefloatConfigXmlChunk> {
    let cmdOffset: Int
    switch commandOffset(payload, expectedCommand: COMM_GET_CUSTOM_CONFIG_XML) {
    case .success(let value): cmdOffset = value
    case .failure(let message): return .failure(message)
    }
    if payload.count < cmdOffset + 10 {
      return .failure("Short Refloat config XML response: \(payload.count - cmdOffset) bytes")
    }
    let confInd = Int(payload[cmdOffset + 1])
    if confInd != expectedConfInd {
      return .failure("Unexpected Refloat config XML index \(confInd)")
    }
    let totalLength = readInt32(payload, cmdOffset + 2)
    let dataOffset = readInt32(payload, cmdOffset + 6)
    if totalLength < 0 { return .failure("Negative Refloat config XML length \(totalLength)") }
    if dataOffset < 0 || dataOffset > totalLength {
      return .failure("Invalid Refloat config XML offset \(dataOffset) for length \(totalLength)")
    }
    let chunk = Array(payload[(cmdOffset + 10)..<payload.count])
    if dataOffset + chunk.count > totalLength {
      return .failure(
        "Refloat config XML chunk exceeds length: offset=\(dataOffset) chunk=\(chunk.count) length=\(totalLength)"
      )
    }
    return .success(RefloatConfigXmlChunk(confInd: confInd, totalLength: totalLength, offset: dataOffset, chunk: chunk))
  }

  static func parseCustomConfigResponse(
    _ payload: [UInt8],
    expectedConfInd: Int = 0
  ) -> RefloatConfigProtocolResult<RefloatConfigBytes> {
    let offset: Int
    switch commandOffset(payload, expectedCommand: COMM_GET_CUSTOM_CONFIG) {
    case .success(let value): offset = value
    case .failure(let message): return .failure(message)
    }
    if payload.count < offset + 6 {
      return .failure("Short Refloat config response: \(payload.count - offset) bytes")
    }
    let confInd = Int(payload[offset + 1])
    if confInd != expectedConfInd {
      return .failure("Unexpected Refloat config index \(confInd)")
    }
    return .success(
      RefloatConfigBytes(
        confInd: confInd,
        packageSignature: UInt32(bitPattern: Int32(readInt32(payload, offset + 2))),
        config: Array(payload[(offset + 6)..<payload.count])
      )
    )
  }

  static func parseGetInfoResponse(_ payload: [UInt8]) -> RefloatConfigProtocolResult<RefloatPackageInfo> {
    let offset: Int
    switch appCommandOffset(payload, expectedCommand: REFLOAT_GET_INFO) {
    case .success(let value): offset = value
    case .failure(let message): return .failure(message)
    }
    let dataOffset = offset + 1
    if payload.count <= dataOffset { return .failure("Short Refloat info response: 0 bytes") }
    if payload[dataOffset] == 2 { return parseGetInfoV2(payload, dataOffset) }
    return parseGetInfoV1(payload, dataOffset)
  }

  private static func commandOffset(
    _ payload: [UInt8],
    expectedCommand: Int
  ) -> RefloatConfigProtocolResult<Int> {
    guard !payload.isEmpty else { return .failure("Empty Refloat config response") }
    let cmd = Int(payload[0])
    if cmd == expectedCommand { return .success(0) }
    if cmd == COMM_FORWARD_CAN {
      if payload.count < 3 { return .failure("Short forwarded Refloat config response") }
      let forwarded = Int(payload[2])
      if forwarded == expectedCommand { return .success(2) }
      return .failure("Unexpected forwarded Refloat config command \(forwarded), expected \(expectedCommand)")
    }
    return .failure("Unexpected Refloat config command \(cmd), expected \(expectedCommand)")
  }

  private static func appCommandOffset(
    _ payload: [UInt8],
    expectedCommand: Int
  ) -> RefloatConfigProtocolResult<Int> {
    if payload.count < 3 { return .failure("Short Refloat app response") }
    let cmd = Int(payload[0])
    if cmd == COMM_CUSTOM_APP_DATA {
      let magic = Int(payload[1])
      let appCommand = Int(payload[2])
      if magic != REFLOAT_MAGIC { return .failure("Unexpected Refloat magic \(magic)") }
      if appCommand == expectedCommand { return .success(2) }
      return .failure("Unexpected Refloat app command \(appCommand), expected \(expectedCommand)")
    }
    if cmd == COMM_FORWARD_CAN {
      if payload.count < 5 { return .failure("Short forwarded Refloat app response") }
      let forwarded = Int(payload[2])
      let magic = Int(payload[3])
      let appCommand = Int(payload[4])
      if forwarded != COMM_CUSTOM_APP_DATA {
        return .failure("Unexpected forwarded Refloat command \(forwarded), expected \(COMM_CUSTOM_APP_DATA)")
      }
      if magic != REFLOAT_MAGIC { return .failure("Unexpected Refloat magic \(magic)") }
      if appCommand == expectedCommand { return .success(4) }
      return .failure("Unexpected forwarded Refloat app command \(appCommand), expected \(expectedCommand)")
    }
    return .failure("Unexpected Refloat app response command \(cmd), expected \(COMM_CUSTOM_APP_DATA)")
  }

  private static func parseGetInfoV1(
    _ payload: [UInt8],
    _ dataOffset: Int
  ) -> RefloatConfigProtocolResult<RefloatPackageInfo> {
    if payload.count < dataOffset + 3 {
      return .failure("Short Refloat info v1 response: \(payload.count - dataOffset) bytes")
    }
    let versionCode = Int(payload[dataOffset])
    return .success(RefloatPackageInfo(version: "Refloat \(versionCode / 10).\(versionCode % 10)"))
  }

  private static func parseGetInfoV2(
    _ payload: [UInt8],
    _ dataOffset: Int
  ) -> RefloatConfigProtocolResult<RefloatPackageInfo> {
    let minLength = dataOffset + 2 + 20 + 3
    if payload.count < minLength {
      return .failure("Short Refloat info v2 response: \(payload.count - dataOffset) bytes")
    }
    let packageName = displayName(fixedString(payload, dataOffset + 2, 20))
    let major = Int(payload[dataOffset + 22])
    let minor = Int(payload[dataOffset + 23])
    let patch = Int(payload[dataOffset + 24])
    let suffix = payload.count >= dataOffset + 45 ? fixedString(payload, dataOffset + 25, 20) : ""
    let suffixPart = suffix.isEmpty ? "" : (suffix.hasPrefix("-") ? suffix : "-\(suffix)")
    return .success(RefloatPackageInfo(version: "\(packageName) \(major).\(minor).\(patch)\(suffixPart)"))
  }

  private static func fixedString(_ payload: [UInt8], _ offset: Int, _ length: Int) -> String {
    let limit = min(offset + length, payload.count)
    var end = offset
    while end < limit && payload[end] != 0 { end += 1 }
    if end <= offset { return "" }
    return String(bytes: payload[offset..<end], encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
  }

  private static func displayName(_ raw: String) -> String {
    if raw.isEmpty { return "Refloat" }
    return raw.prefix(1).uppercased() + raw.dropFirst()
  }

  private static func int32Bytes(_ value: Int) -> [UInt8] {
    [
      UInt8((value >> 24) & 0xff),
      UInt8((value >> 16) & 0xff),
      UInt8((value >> 8) & 0xff),
      UInt8(value & 0xff),
    ]
  }

  private static func readInt32(_ bytes: [UInt8], _ offset: Int) -> Int {
    let value = (UInt32(bytes[offset]) << 24)
      | (UInt32(bytes[offset + 1]) << 16)
      | (UInt32(bytes[offset + 2]) << 8)
      | UInt32(bytes[offset + 3])
    return Int(Int32(bitPattern: value))
  }
}
